src/Nri/Ui/Button/V9.elm

```diff
@@ -0,0 +1,887 @@
+custom attribute (ButtonOrLink config) =
+    ButtonOrLink { config | customAttributes = attribute :: config.customAttributes }
+
+
+
+-- LINKING, CLICKING, and TRACKING BEHAVIOR
+
+
+{-| -}
+onClick : msg -> Attribute msg
+onClick msg (ButtonOrLink config) =
+    ButtonOrLink
+        { onClick = Just msg
+        , url = config.url
+        , linkType = config.linkType
+        , size = config.size
+        , style = config.style
+        , width = config.width
+        , label = config.label
+        , state = config.state
+        , icon = config.icon
+        , customAttributes = config.customAttributes
+        }
+
+
+{-| -}
+href : String -> Attribute msg
+href url (ButtonOrLink config) =
+    ButtonOrLink
+        { onClick = config.onClick
+        , url = url
+        , linkType = config.linkType
+        , size = config.size
+        , style = config.style
+        , width = config.width
+        , label = config.label
+        , state = config.state
+        , icon = config.icon
+        , customAttributes = config.customAttributes
+        }
+
+
+type Link
+    = Default
+    | WithTracking
+    | SinglePageApp
+    | WithMethod String
+    | External
+    | ExternalWithTracking
+
+
+{-| Use this link for routing within a single page app.
+
+This will make a normal <a> tag, but change the Events.onClick behavior to avoid reloading the page.
+
+See <https://github.com/elm-lang/html/issues/110> for details on this implementation.
+
+-}
+linkSpa : Attribute msg
+linkSpa buttonOrLink =
+    buttonOrLink
+
+
+{-| Wrap some text so it looks like a button, but actually is wrapped in an anchor to
+some url, and it's an HTTP request (Rails includes JS to make this use the given HTTP method)
+-}
+linkWithMethod : String -> Attribute msg
+linkWithMethod method (ButtonOrLink config) =
+    ButtonOrLink { config | linkType = WithMethod method }
+
+
+{-| Wrap some text so it looks like a button, but actually is wrapped in an anchor to some url.
+This should only take in messages that result in a Msg that triggers Analytics.trackAndRedirect. For buttons that trigger other effects on the page, please use Nri.Button.button instead
+-}
+linkWithTracking : Attribute msg
+linkWithTracking (ButtonOrLink config) =
+    ButtonOrLink { config | linkType = WithTracking }
+
+
+{-| Wrap some text so it looks like a button, but actually is wrapped in an anchor to
+some url and have it open to an external site
+-}
+linkExternal : Attribute msg
+linkExternal (ButtonOrLink config) =
+    ButtonOrLink { config | linkType = External }
+
+
+{-| Wrap some text so it looks like a button, but actually is wrapped in an anchor to some url and have it open to an external site
+
+This should only take in messages that result in tracking events. For buttons that trigger other effects on the page, please use Nri.Ui.Button.V2.button instead
+
+-}
+linkExternalWithTracking : Attribute msg
+linkExternalWithTracking (ButtonOrLink config) =
```

Comment:
🎨 for some of these extra link attributes, I'm not sure I understand what all are used for, but I think some for instance will only be used when there's also an href and/or a msg -- in cases like those, I think it'd be nice to have the attribute function handle multiple private options at once, like `linkExternalWithTracking : Url -> msg -> Attribute msg` (which sets onClick, href, and linkType all at once)

Reply:
I thought this was a great suggestion and I started it, but I ran into this:

```
linkWithMethod : String -> String -> Attribute msg
linkWithMethod method url (ButtonOrLink config) =
    ButtonOrLink { config | linkType = WithMethod method, url = url }
```

We don't have the Url type defined in noredink-ui. I wouldn't want the Url data type to be versioned/seems like a pain to add it.

I guess I could make this a record? What do you think?

Reply:
Hmm, yeah, I guess a record would be fine then.  Also I think just having two strings and documenting the order of them in the doc comment would also probably be fine 🤷‍♀

Reply:
All done!